test/saithrift/tests/SAITest.md

```diff
@@ -0,0 +1,122 @@
+Overview   
+========
+
+|  Testcase ID 	|         Test Catagory	 		| Number of TCs |   Owned by	|   Comments	|
+|---		|---					|---		|---		|---		|
+|   	1-2	|  L3 Directed Broadcast		|   2		|   		|   		|
+|   	3	|  L3 Loopback subnet			|   1		|   		|   		|
+|   	4-5	|  L2/L3 MTU				|   2		|   		|   		|
+|   	6-8	|  IPv4 /32 route and same Neighbor IP 	|   3		|   		|   		|
+|   	9	|  Dual/Triple tagging 			|   1		|   		|   		|
+|   	10-12	|  MAC move 				|   3		|   		|   		|
+|   	13-16	|  Neighbor MAC change and age-out 	|   4		|   		|   		|
+|   	17-19	|  Traffic/FDB learning on Bridgeport 	|   3		|   		|   		|
+|   	20-21	|  LAG/ECMP Hash seed 			|   2		|   		|   		|
+|   	22-26	|  IPv4/v6 ECMP Group member 		|   5		|   		|   		|
+|   	27-32	|  ACL Table/Group and bindings 	|   6		|   		|   		|
+
+List of Extended Test Cases
+===========================
+
+1.	L3DirectedBroadcast - I  
+Send IPv4 packet to a broadcast IP. The packet must have unicast destination MAC (Gateway MAC) and a broadcast destination IP (like 192.168.0.255). Verify that the packet is sent out on all member ports of VLAN if the outgoing interface is a VLAN RIF. 
+
+2.	L3DirectedBroadcast – II (VLAN checks)  
+After executing “L3DirectedBroadcast – I”, add and remove ports to the VLAN.  When a port is added, the L3 broadcast traffic must also be forwarded on the newly added member port and similar test to be executed by removing existing member port from VLAN. 
+
+3.	L3LpbkSubnetTest  
+Send traffic on a layer 3 router interface destined to an IPv4 address within the same subnet. Have the nhop entry learnt on the same interface as incoming interface. Verify that packet is send out on the same outgoing interface as the incoming interface. Set the SAI_ROUTER_INTERFACE_ATTR_LOOPBACK_PACKET_ACTION attribute to _drop_ and verify that packet is _not_ send out on the same outgoing RIF as incoming RIF. 
+
+4.	L3MtuTest  
+Send L3 destined packet with max interface MTU and verify forwarding. With L3 destined packet, verifying the forwarding behavior also ensures IP MTU value configured for the router interface. Increase the packet size beyond the configured MTU value and verify the behavior. Default action is to drop packets exceeding configured MTU
+
+5.	L2MtuTest  
+Send untagged packet with max MTU size and verify forwarding.  Add VLAN tag so that the total packet size exceeds the MTU value and verify behavior. Default action is to drop packets exceeding configured MTU value. 
+
+6.	L3IPv4/32Test - I  
+(Ref: https://github.com/opencomputeproject/SAI/blob/master/doc/behavioral%20model/pipeline_v6.pdf, router flow)
+Add a route entry with /32 prefix (e.g. 10.1.1.1/32 via NH1) and a neighbor entry for the same IP address (e.g. 10.1.1.1 MAC1 Port1). Send traffic and verify that the packet takes the path specified by the /32 route entry and _not_ the neighbor path. 
```

Comment:
For the directly reachable routes, egress_rif will be filled in the SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID attribute.
And, the "egress_rif, pkt_dst_ip" would be matched in the "Neigh table" instead of the "egress_rif,NH_DstIP" as captured in the "Ingress router flow" diagram.
In that case, there will be two matches for the packet in this scenario (/32 route && neighbor entry coexist) - both in the Router table and in the Neigh table. Then, there should be a stage that determines which one of the matches will take precedence. It looks like that is not defined explicitly in the pipeline model currently. 
@lguohan @atitjain Comments?

Reply:
@SathishKumarTR-Dell , the packet should first hit the route table, then it will hit neigh table, there is no precedence between route table and neighbor table.